### PR TITLE
Fix: Unhandled Errors thrown inside a .then() handler in getAccessToken

### LINKED ISSUE #1134: Fix unhandled rejection in getAccessToken
This PR needs to be redone:
https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1090

to add changes to `localStorage.native.js` as well.

The new PR needs to be properly tested.

Possibly add jest tests.

### PR DESCRIPTION
**Pull request by crucialfelix**
This fixes minor issues in `app/utils/user.js` `getAccessToken` (used for logging in and session management)

There are many reported errors coming from `/login`:

https://app.bugsnag.com/plant-for-the-planet/plant-for-the-planet-app/errors/5d05f81ab3cb100019330da9?filters[search][0][type]=eq&filters[search][0][value]=UnhandledRejection

```
3 hours ago Aug 2nd, 20:48:44 UTC 
UnhandledRejection/login
Rejection reason was not an Error. See "Promise" tab for more detail.
rejection reason
undefined (or null)
(No stack trace)
```

This PR doesn't attempt to fix the error - it is just to supply an Error and get a stack trace so we know what the error is so it can then be fixed. 

Could be:
- localStorage failure (web or react-native), 
- API post request timeout.
- Unknown unknown

It was previously submitted and merged: https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1090 

This was missing one function in `localStorage.native.js`

@norbertschuler reverted the commit in: https://github.com/Plant-for-the-Planet-org/treecounter-app/commit/ceae9845197883ac6d5b3718e0c6b0979138951d
and then there were several other attempts (while I was on vacation) to revert, fix, reapply etc.

eg. https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1133
which was merged, then again mostly reverted.

For a minor fix, it's now become quite confusing and time consuming.

The original PR fixed:

1. Always reject with an Error — this is how bugsnag gets it's stack trace.
  That fix is now already in develop.

2. Flattened the mixed pyramids of Promise and async functions so that it's easier to follow, test and debug.

---

It is newly rebased on develop.

I tested login / logout and session refresh on web and react-native iOS (same code as for Android)

Fixes #1134  - Fix unhandled rejection in getAccessToken

### CODE REVIEW

**Comment by norbertschuler** [comment]
@niteshgrg please check this pull request and decide if you'd prefer it to your own pull request https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1133 ?

**Comment by norbertschuler** [comment]
@crucialfelix @niteshgrg 
I first merged https://github.com/Plant-for-the-Planet-org/treecounter-app/pull/1133 but reverted the commits again now after I found out staging can not react correctly on an outdated refresh-token which results into the refresh-API call `POST https://staging.trilliontreecampaign.org/api/token/refresh?version=v1.1&_locale=de` (plus a JSON payload with the refresh token inside) to fail with

```
{"code":400,"message":"Validation Failed","errors":{"errors":["Bad credentials"],"children":{"_username":{"errors":["Bad credentials"]},"_password":[]}}}
```
It also fails with the following exception if I delete the refresh-token from the local database of the browser:
```
localStorage.js:30 Uncaught (in promise) Error: token not in localStorage
    at localStorage.js:30
    at new Promise (<anonymous>)
    at t.fetchItem (localStorage.js:27)
    at user.js:10
    at tryCatch (runtime.js:65)
    at Generator._invoke (runtime.js:303)
    at Generator.e.<computed> [as next] (runtime.js:117)
    at step (bundle.js:50111)
    at bundle.js:50111
    at new Promise (<anonymous>)
```

If this patch can be made the way it's avoid above problems, that would be nice.

**Comment by norbertschuler** [comment]
As login on staging server was broken, I felt forced to make a fix in the develop branch by these three commits which now conflicts with your branch: 
- aee3e41fa7ec27657e0a078a70128c9a19e845c0
- dee53f05d1361a23af8b1c18fd61df2281117fc3 (this one is just a stupid error changed by the next commit)
- ecac5a7b213c1b42665b05486e91ad7f795f0143

**Comment by norbertschuler** [comment]
@crucialfelix maybe we now do not need this pull request any more. Can you check?

**Comment by norbertschuler** [comment]
I will check this once more after the release 1.0.5 is done and merge if it works. I am just unsure if all rejections from promised will be caught and the code not stopped by the rejections as of now.

File: app/utils/user.js [hunk 1]
```diff
@@ -7,16 +7,21 @@ import { postRequest } from './api';
...
+      // This may throw an Error if localStorage is broken
+      // or POST requests timeout
+      const newToken = await refreshTokenIfExpired();
+      if (newToken) {
+        return newToken;
+      }
     }
+    return token;
+  } catch (err) {
+    console.log(err);
```
Comment (crucialfelix) [inline app/utils/user.js]: > the app shouldn't just stop on a 400 error and throw an exception, but continue to show the login

So an error happened in `refreshTokenIfExpired()` — a 400 error returned from the server. That's the problem. That problem needs to be handled there, not here.

The declared responsibility of `refreshTokenIfExpired` is to do just that "refresh token if expired". It must handle errors.

By catching it here and merely logging it we will no longer have any idea that there is a problem, and the user is in some weird state.

File: app/utils/user.js [hunk 2]
```diff
@@ -47,6 +41,22 @@ const getExpirationTimeStamp = token => {
   return getCurrentUnixTimestamp() + exp - iat;
 };
 
+/**
+ * @returns string | void
+ */
+const refreshTokenIfExpired = async () => {
+  if (await tokenIsExpired()) {
+    const prev_refresh_token = await fetchItem('refresh_token');
+    const response = await postRequest('gesdinet_jwt_refresh_token', {
```
Comment (crucialfelix) [inline app/utils/user.js]: This code is copied from the original, just flattened a bit.  It seems the problem is here then. That's what was raising all those errors.

We need to check the status of response and then the function needs to do the right thing.

400 ? So the user is not authorized?  That means Bad Request.  What was bad about the request? What is `prev_refresh_token` at this point for this Bad User?

Or was this a kind of 500 error (server bug) that returns a 400?

Regardless, the server response should be detected here and should probably return a null token. try catch is not the right way to write business logic, and that's kind of why this code is causing confusion and errors.

**Comment by crucialfelix** [comment]
So what was the 400 error all about? Can I get logs from the server?

**Comment by crucialfelix** [comment]
And how did you manage to trigger this error? I tested it with a legit login, then I edited the code to make the token expired so it would refresh and get a new token.

**Comment by norbertschuler** [comment]
I am sorry, I currently do not know how I managed to get the 400 bad request error. I will have to examine this to find out how to reproduce it. Of course you're right, it's not a good solution to have the try/catch inserted - it's just otherwise it won't work and shows a blank page. If you could rewrite it to catch the error where the `postRequest` failed the way it does not stop code execution so the user can be shown the login page again, that would be great. Feel free to revert my commit with the try/catch block again.

**Comment by crucialfelix** [comment]
Will do. btw. I find it easier to just get code review comments (which are always helpful). Adding commits confuses the development flow and adds maintenance time. Thanks

**Comment by norbertschuler** [comment]
I think I know how to reproduce the 400 request. I edited my local storage in Chrome by changes the value of `token_expires` to 0 and also the value of `refresh_token` to 0. Then the following request returned the result below:

Request:
`POST https://www.trilliontreecampaign.org/api/token/refresh?version=v1.1&_locale=de`
with payload 
`{"refresh_token":"0"}`

Result:
```json
{"code":400,"message":"Validation Failed","errors":{"errors":["Bad credentials"],"children":{"_username":{"errors":["Bad credentials"]},"_password":[]}}}
```

**Comment by crucialfelix** [comment]
I did the same kind of test, but I guess my token was still active. Yours was actually dead. 400 may be valid. Jorge will confirm.

I'll check response in refresh and then it's legit to return null. (not an error)

**Comment by crucialfelix** [comment]
Why is it returning `{"_username":{"errors":["Bad credentials"]},"_password":[]}` for refresh token? That seems like the wrong response.

**Comment by norbertschuler** [comment]
> Why is it returning `{"_username":{"errors":["Bad credentials"]},"_password":[]}` for refresh token? That seems like the wrong response.

That's a question I cannot answer as I have build the API. Probably something to optimize.

**Comment by norbertschuler** [comment]
@crucialfelix Can I do anything to help you get this ready to merge?

**Comment by crucialfelix** [comment]
Should be ready now. I moved the try catch to directly around the api call.

**Comment by norbertschuler** [comment]
O.k., now if the `token_expires` value triggers a refresh and the `refresh_token` is invalid, the call to refresh the token `gesdinet_jwt_refresh_token` fails, but as long as the `token` is still valid, I am still logged in (before I was requested to login again). But I think that's o.k. - you just have errors caught in the browser until you logout and login again to refresh all this local variables.
The only thing bugging me is the error #1239 which occurred as I first checked out this branch and opened by Chrome browser. But I missed to investigate it and could not reproduce it now any more.
So let's merge this.

**Comment by crucialfelix** [comment]
Re #1239 
You got a "Error · Request failed with status code 401" which is the correct response from the server: not authenticated.

All calls got getRequest or getAuthenticatedRequest get checked:
https://github.com/Plant-for-the-Planet-org/treecounter-app/blob/2faff94a34801ef31958a2d4e96e70ad183f2175/app/utils/api.js#L21-L33

So I think that is correct behavior.

**Pull request merged**